### ISSUE TITLE
A question about how PaddlePaddle implements the L2 regularization.

### ISSUE THREAD

**Issue by lcy-seso** [PaddlePaddle/Paddle#5899]
As far as I understand, after one forward-backward computation, when updating the learnable weights, the order is:

1. Perform gradients clipping (if needed, including hard clipping, soft clipping, clipping by norm, and so on)
2. Apply regularization (if needed). The regularization terms express our prior beliefs about the solution.
3. Calculate the adaptive gradients according to different formula of the different optimization algorithms, such as Adagrad, Rmsprop, Adadelta, Adam, Adammax, and so on.

This is because the cost is usually made up of two parts:

1. The error measure from a particular loss function.
2. The penalty terms to avoid overfitting (the regularization term).

As a result, the gradient of the loss with respect to a learnable parameter is also made up of two parts: (1) the gradient from the loss function; (2) the loss from the regularization term.

When applying the adaptive gradient algorithms, we should sum up the gradients from the regularization term and the gradients from the loss function first.

In PaddlePaddle's codes, here I take the Adagrad for example, regularization is applied after performing the adaptive gradient algorithm, which affects algorithm like Adam when estimating the momentum.

https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/math/TrainingAlgorithmOp.cu#L80

```cpp
void adagradApply(BaseMatrix& value,
                  BaseMatrix& grad,
                  BaseMatrix& mom,
                  BaseMatrix& accum_buffer,
                  BaseMatrix& accum,
                  BaseMatrix& lr,
                  real epsilon,
                  real learningRate,
                  real momentum,
                  real decayRate) {
  auto expr1 = accum.lazyAssign(accum + grad.square());
  auto expr2 =
      lr.lazyAssign((accum_buffer + accum + epsilon).sqrt().reciprocal());
  auto expr3 = mom.lazyAssign(mom * momentum -
                              learningRate * lr * (grad + value * decayRate));
  auto expr4 = value.lazyAssign(value + mom);

  AssignEvaluate(expr1, expr2, expr3, expr4);
}
```
A very interesting thing is, decoupling weight decay and the optimization steps are proposed by a recent paper: [Fixing Weight Decay Regularization in Adam](https://arxiv.org/abs/1711.05101), and is claimed to have a better learning performance on CIFAR-10 and ImageNet32x32. The green color one is the way PaddlePaddle uses and the variant proposed by the paper.

![image](https://user-images.githubusercontent.com/5842774/33203932-b99636b4-d13c-11e7-98f9-0ba08fc57fb2.png)

@emailweixu I am very interested in (and curious about) the way PaddlePaddle applies the L2 regularization. I am not very sure whether my understanding about PaddlePaddle is right and why we implement the L2 regularization in this way before. Thank you for the help.

**Comment by emailweixu** [comment]
Basically, in PaddlePaddle's implementation, the magnitude of a parameter does not affect its learning rate (through weight decay).   Only the gradient affects the its learning rate.

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
